crates/driver/src/infra/api/routes/solve/dto/auction.rs

```diff
@@ -64,24 +64,34 @@ impl Auction {
                         } else {
                             competition::order::Partial::No
                         },
-                        pre_interactions: order
-                            .pre_interactions
-                            .into_iter()
-                            .map(|interaction| eth::Interaction {
-                                target: interaction.target.into(),
-                                value: interaction.value.into(),
-                                call_data: interaction.call_data.into(),
-                            })
-                            .collect(),
-                        post_interactions: order
-                            .post_interactions
-                            .into_iter()
-                            .map(|interaction| eth::Interaction {
-                                target: interaction.target.into(),
-                                value: interaction.value.into(),
-                                call_data: interaction.call_data.into(),
-                            })
-                            .collect(),
+                        pre_interactions: match order.executed.is_zero() {
+                            true => order
+                                .pre_interactions
+                                .into_iter()
+                                .map(|interaction| eth::Interaction {
+                                    target: interaction.target.into(),
+                                    value: interaction.value.into(),
+                                    call_data: interaction.call_data.into(),
+                                })
+                                .collect(),
+                            // custom interactions of partially fillable orders should only be
+                            // executed on the first fill
+                            false => vec![],
+                        },
+                        post_interactions: match order.executed.is_zero() {
```

Comment:
I can see how pre-interactions may only have to be executed on the first fill. However, for post interactions, I'm not sure this is so clear. The use case we envision for post interactions are things like bridging, staking, etc which could happen after every fill and if only once, they should happen on the last fill (which is impossible to predict)

Reply:
Hmm, that's a good point. I guess we would only know the last fill for sure when encoding the solution (executed_amount_after_fill == total_amount) and only then make that part of the solution. On the other hand this would mess with the fee computation as we would somehow have to make solvers aware that there are additional costs for filling completely.
Also AFAIK some solvers currently have the problem that they leave dust in partially fillable orders meaning that it's uneconomic for other solvers to fill the remaining dust to actually trigger the post interaction.

Reply:
We can change the PR to allow pre-interactions for all order types (but keep post-interactions disabled for partially fillable orders)?

Reply:
I now went for pre-interactions only on first fill and post-interactions on every fill.
Because there is currently no mechanism to tell an interaction how much of an order will be filled the hook basically has to prepare for the "worst case" from the perspective of the hook.
For pre-interactions that means it has to assume it's a 100% fill and needs to produce all the `sell_token` to support the whole order. It only makes sense to do that once.

For post-interactions the situation is not so clear cut. IMO the only sensible options are (1) execute after last fill or (2) execute after every fill. Because some edge cases might result in an order not being filled 100% I think we should opt for `2`. Additionally from the perspective of a hook implementer writing a hook compatible with `2` is the safest option and since we don't have any on-chain guarantees for the execution of hooks implementers should prepare for the worst to be safe.
AFAICS the only consequence of this is that the hook should be written in a way that it checks the fill status of the order to determine the correct action.
So in conclusion although I initially thought there are multiple viable approaches for executing interactions for partial fills I now believe there is only 1 reasonable decision here which means we don't have to add any sophisticated mechanism to let the user decide how/when their interactions get executed (can't be enforced anyway so why bother).

Reply:
Until we have people demanding for it, I wouldn't do it. But would we want to enhance the semantics on how we define pre/post hooks to give an indication on how to execute this? 

This can be just a soft contract, the protocol don't enforce it, but is still nice to have this flexibility.

In principle, post-interactions could be even done in another transaction. So one potential solution to post-hooks for partial fills, is to do it in the next block (or next settlement) once we know the order is 100% filled.  I don't think this is good in every case, but its possibly a construct that is useful for many cases (like TWAP and bridge at the end).

Reply:
>Until we have people demanding for it, I wouldn't do it.

Why? We are on it right now and I follow Martin's logic that there is only one good way to implement it, so why not include it? When someone needs it we need to make ressources for it, switch context and accompany the change in production which will take time.

> So one potential solution to post-hooks for partial fills, is to do it in the next block

I don't think there is any value in delaying post hook execution into the next block (as it means it could potentially frontrun by other actors). By the time we craft the settlement tx, we know as much as we know in the next block so we may as well include it right away. In particular we won't know, if an order is 99% filled, if it will ever get another fill or not (may depend on gas and price swings) so we can never really decide when the last fill happened.

Reply:
> Why? We are on it right now and I follow Martin's logic that there is only one good way to implement it, so why not include it? When someone needs it we need to make ressources for it, switch context and accompany the change in production which will take time.

I think there's a misunderstanding. I can see my message was a bit misleading. I agree with the proposed solution, what I want to say, is that we might want to enhance the `appData` definition. More specifically we can add more information in the Hook definition on how the user wants to execute the pre/post interactions: see https://github.com/cowprotocol/app-data/blob/main/src/schemas/hook/v0.1.0.json

With "Until we have people demanding for it," i was referring to doing this enhancement to allow users to be more specific on how they want their hooks executed. Martin's proposal makes a great default, and I would love to see it implemented

Reply:
> I don't think there is any value in delaying post hook execution into the next block (as it means it could potentially frontrun by other actors). By the time we craft the settlement tx, we know as much as we know in the next block so we may as well include it right away. In particular we won't know, if an order is 99% filled, if it will ever get another fill or not (may depend on gas and price swings) so we can never really decide when the last fill happened.

For me the posthook default makes sense when is 100%. We don't know if it will ever be executed if there's some 1% missing, but we know when the order expires. 
I think is a fair pragmatic default to:
- Execute only when is 100% (if dust is a problem we might want to allow a tiny error tolerance threshold)
- Also execute if the order is expired and partially filled